### DISCUSSION TITLE
General Issues with Artworks database table with forms and tables.

### DISCUSSION THREAD

**Discussion by TheFrankman** [filamentphp/filament#9932] [Help] [answered]
I'm trying to use the FileUpload on forms, as well as ImageColumn on tables etc, but utilising an artworks table for the various artworks.

The reason for this being that an artwork can be used on a release, or an artist. An artist has one artwork, a release has one artwork. The classes below to better illustrate.

The artwork table has a `name`, and a `path` column. The path is the relative path on the disk.
The releases table has an `artwork_id` column
The artists table has an `artwork_id` column

```
class Release extends Model
{
...
    public function artwork(): BelongsTo
    {
        return $this->belongsTo(Artwork::class);
    }
}
```

```
class Artist extends Model
{
...
    public function artwork(): BelongsTo
    {
        return $this->belongsTo(Artwork::class, 'artwork_id');
    }
}
```

```
class Artwork extends Model
{
...
    public function releases(): BelongsToMany
    {
        return $this->belongsToMany(Release::class);
    }
   
   public function artists(): BelongsToMany
    {
        return $this->belongsToMany(Artist::class);
    } 
}
```

There are a couple of issues I am having with this implementation but let's start with the most simple one. On my ArtistResource i'd like to display the artwork in the database column : 

```
 public static function table(Table $table): Table
{
    return $table
            ->columns([
                ImageColumn::make('artwork'),
                Tables\Columns\TextColumn::make('name')
                    ->searchable(),
                Tables\Columns\TextColumn::make('legal_first_name')
                    ->searchable(),
}    
```

I'm given to understand that this should work provided that artwork has a path attribute. I have also tried  ` ImageColumn::make('artwork.path'),` and modifying my `Artist` class to have `protected $with = ['artwork']` to ensure the relationship is eager loaded.

The result is no errors, but an empty column.

I am also having all sorts of issues with the form side of stuff which is a bit to convoluted to explain here, but perhaps someone could share their experience with storing their media in this way. I have looked at the demo project and it seems that they use the Spatie Media Manager to achieve these results, but i feel it should be possible out of the box.

**Comment by smiliyas** [comment]
Try this

```PHP
ImageColumn::make('artwork.path')->formatStateUsing(fn ($state) => $state ? url($state) : null);
```

**Reply by TheFrankman** [reply]
formatStateUsing is not a function

**Reply by smiliyas** [reply]
Might be formatState() like that

**Reply by TheFrankman** [reply]
state() is, but it causes an infinite loop.

**Reply by TheFrankman** [reply]
also there is no format* method of any kind for this column type.

**Reply by smiliyas** [reply]
is anything like formatUsing() there?

**Reply by smiliyas** [reply]
Try with getStateUsing()

**Reply by smiliyas** [reply]
Like this

```PHP
ImageColumn::make('artwork.path')->getStateUsing(fn ($state) => $state ? url($state) : null);
```

**Reply by TheFrankman** [reply, marked as answer]
This is what i managed to get working : 

```
ImageColumn::make('artwork')
                    ->state(function (Model $artist): ?string {
                        return $artist->artwork?->path;
                    }),
```
